### PR TITLE
Class Implemented

### PR DESCRIPTION
**Pull request by evan-fannin**
### Description

This PR introduces the `LLMFileAccessController` class, which implements support for reading a `.clineignore` file to validate file access for the LLM. The following are features:

- Support for `.clineignore` files using familiar .gitignore syntax, allowing users to easily specify which files to exclude
- Default security patterns that protect sensitive files (`.env`, `.git`, credentials, etc.)
- Path normalization and security checks to prevent directory traversal
- Batch filtering capability for efficiently processing multiple paths
- Error handling with a fail-closed approach

### Test Procedure

The implementation is tested with 20+ test cases covering:
- Default security patterns (blocking sensitive files)
- Custom pattern handling from .clineignore
- Path normalization and security (absolute/relative paths, directory traversal)
- Edge cases (empty files, invalid paths, missing .clineignore)
- Batch filtering functionality

### Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 📚 Documentation update

### Pre-flight Checklist

- [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
- [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
- [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Additional Notes

- The implementation uses the 'ignore' library to support standard .gitignore syntax, making it intuitive for users familiar with Git
- A TODO has been added for support of negation patterns (e.g., `!temp/allowed/*`)

----

> [!IMPORTANT]
> Introduces `LLMFileAccessController` to manage file access using `.clineignore` files with comprehensive testing and error handling.
> 
>   - **New Class**:
>     - `LLMFileAccessController` in `LLMFileAccessController.ts` to manage file access using `.clineignore` files.
>     - Supports `.clineignore` syntax similar to `.gitignore` for specifying file access rules.
>     - Implements path normalization to prevent directory traversal.
>     - Provides batch filtering for multiple paths.
>     - Includes error handling with a fail-closed approach.
>   - **Testing**:
>     - `LLMFileAccessController.test.ts` includes 20+ test cases.
>     - Tests cover default security patterns, custom patterns, path normalization, edge cases, and batch filtering.
>   - **Dependencies**:
>     - Adds `ignore` library to `package.json` for handling ignore patterns.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 41501a72e4ec42245e6622ec6202c53ac07a05c8. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by evan-fannin** [comment]
Need to fix the ignore library to handle negation patterns better. Currently it's returning negative as soon as it encounters a parent directory that's been negated, not following the .gitignore spec.

File: src/services/llm-access-control/LLMFileAccessController.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		"**/build/dependencies/**", // Dependencies within build folders
+		"**/.*", // Hidden directories
+	]
+
+	constructor(cwd: string) {
+		this.cwd = cwd
+		this.ignoreInstance = ignore()
+
+		// Add default patterns immediately
+		this.ignoreInstance.add(LLMFileAccessController.DEFAULT_PATTERNS)
```
Comment (saoudrizwan) [inline src/services/llm-access-control/LLMFileAccessController.ts:25]: I dont think we should be always ignoring these default patterns. I've needed cline to edit my env files a ton

**Review by dcbartlett — DISMISSED** [review]

File: src/services/llm-access-control/LLMFileAccessController.ts (hunk 1 above)
Comment (dcbartlett) [reply src/services/llm-access-control/LLMFileAccessController.ts:25]: Actually I agree with this on `.env` `/env/` and `/venv/` as these are normally project dependent.  

Should we instead have Cline create a .clineignore with these defaults in it if a .clineignore doesn't exist?  This would make it possible for the end-user to see what the defaults are and change them.

Comment (celestial-vault) [reply src/services/llm-access-control/LLMFileAccessController.ts:25]: Yes we will probably want some granularity where certain files can be accessed in some operations versus others (e.g. writing to a file if explicitly told versus having the file's contents listed in context) @saoudrizwan

Comment (saoudrizwan) [reply src/services/llm-access-control/LLMFileAccessController.ts:25]: lets not overcomplicate this, i think just remove the default patterns to have current behavior. and IF user creates clineignore then we apply this

Comment (celestial-vault) [reply src/services/llm-access-control/LLMFileAccessController.ts:25]: @saoudrizwan Removed some of the defaults, leaving it to patterns I don't envision the vast majority of users wanting touched. We can always edit this and introduce further granularity (for listing versus writing, etc.)

File: src/services/llm-access-control/LLMFileAccessController.ts [hunk 2]
```diff
@@ -43,13 +40,10 @@ export class LLMFileAccessController {
...
 		"**/bundle/**", // Bundled files for deployment
 		"**/vendor/**", // Vendor dependencies (used in PHP, Go, etc.)
-		"**/tmp/**", // Temporary files
-		"**/temp/**", // Temporary files
 		"**/deps/**", // Dependencies folder (various ecosystems)
 		"**/pkg/**", // Package files (common in Go, Rust, etc.)
 		"**/Pods/**", // CocoaPods dependencies (iOS/macOS development)
 		"**/build/dependencies/**", // Dependencies within build folders
-		"**/.*", // Hidden directories
 	]
```
Comment (saoudrizwan) [inline src/services/llm-access-control/LLMFileAccessController.ts:19]: can we pls just remove default apttersn altogether? i envision complication where if this doesnt match with the dirsToIgnore in listFiles, then listFiles may list a bunch of files ignored here with (private file, do not read).

Comment (celestial-vault) [reply src/services/llm-access-control/LLMFileAccessController.ts:19]: @saoudrizwan Understood, will be handling the dirsToIgnore accordingly. Removed for now.

**Review by saoudrizwan — APPROVED** [review]
